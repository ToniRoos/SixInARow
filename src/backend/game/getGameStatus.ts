import { GameState, GameStatus as PlayerStatus } from "../game2";

const getGameStatus = (gameState: GameState, id?: string): PlayerStatus => {
    const players = gameState.userStore.getUserSessions().map(userData => userData.name);
    const userSession = gameState.userStore.getSession(id);

    return {
        players,
        state: gameState.state,
        activeTurn: gameState.activePlayerId === id,
        board: gameState.board?.getBoardData(),
        tilesOnHand: userSession ? userSession.tilesOnHand : [],
        activePlayer: gameState.activePlayerName
    };
};

export { getGameStatus };