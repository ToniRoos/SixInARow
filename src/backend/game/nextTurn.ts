import { unknownPlayerError } from "../createError";
import { GameState } from "../game2";
import { UserSessionData } from "../userStore/UserSessionData";

const nextTurn = (gameState: GameState, id: string) => {
    const sessionData = gameState.userStore.getSession(id);

    if (!sessionData) {
        throw unknownPlayerError();
    }

    const tiles = gameState.stock!.getNextTiles(sessionData.tilesOnTurn.length);
    sessionData.tilesOnTurn = [];

    sessionData.tilesOnHand = [...sessionData.tilesOnHand, ...tiles];

    let nextPlayer: UserSessionData = gameState.userStore.getNextPlayer(id);
    gameState.activePlayerId = nextPlayer.id;
    gameState.activePlayerName = nextPlayer.name;
}

export { nextTurn };