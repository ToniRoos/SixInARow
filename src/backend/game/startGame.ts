import { board } from "../board/board";
import { unknownPlayerError } from "../createError";
import { GameRunningState, GameState } from "../game2";
import { createStock } from "../tileStock/createStock";
import { stockConfig } from "../tileStock/stockConfig";
import { UserSessionData } from "../userStore/UserSessionData";

const startGame = (gameState: GameState, id: string) => {

    const userSession = gameState.userStore.getSession(id);
    if (!userSession) {
        throw unknownPlayerError();
    }

    gameState.stock = createStock();
    gameState.board = board();
    gameState.state = GameRunningState.started;

    const userSessionList: UserSessionData[] = gameState.userStore.getUserSessions();
    userSessionList.forEach(userSession => {

        const initalTilesOnHand = gameState.stock!.getNextTiles(stockConfig.amountOfTiles);
        userSession.tilesOnHand = initalTilesOnHand;
    });

    gameState.activePlayerId = userSession.id;
    gameState.activePlayerName = userSession.name;
};

export { startGame };