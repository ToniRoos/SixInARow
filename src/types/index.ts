export type Commands = 'SetId' |
    'SetName' |
    'GetPlayers' |
    'PlayersList' |
    'StartGame' |
    'GetTiles' |
    'CheckMove' |
    'RefreshBoard' |
    'NextTurn';

export interface Command extends SessionId {
    command: Commands,
    data?: any
}

export interface SessionId {
    id: string
}

export enum ColorEnum {
    blue = 1,
    green = 2,
    orange = 3,
    yellow = 4,
    red = 5,
    purple = 6
}

export enum SymbolEnum {
    square = 1,
    circle = 2,
    star1 = 3,
    hash = 4,
    leaf = 5,
    star2 = 6
}

export interface TileSymbols {
    id?: number;
    color?: ColorEnum;
    symbol?: SymbolEnum;
}

export interface TileData extends TileSymbols {
    x: number;
    y: number;
    allowDrag?: boolean;
}

export interface BoardData {
    tiles: TileData[];
    sizeX: number;
    sizeY: number;
    activePlayer: string;
}

export const color1 = "warning";
export const color2 = "success";