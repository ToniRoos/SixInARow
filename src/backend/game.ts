
import { BoardData, TileData, TileSymbols } from "../types";
import { Board } from "./board/board";
import { checkMove } from "./game/checkMove";
import { getGameStatus } from "./game/getGameStatus";
import { nextTurn } from "./game/nextTurn";
import { startGame } from "./game/startGame";
import { TileStock } from "./tileStock/tileStock";
import { createUserStore } from "./userStore/createUserStore";
import { UserStore } from "./userStore/userStore";
import { tileToString } from "./utils/tileToString";

export const enum GameRunningState {
    init = "INIT",
    started = "STARTED"
}

export interface GameStatus {
    players: string[];
    state: GameRunningState;
    activeTurn: boolean;
    board?: BoardData;
    tilesOnHand: TileSymbols[];
    activePlayer: string;
}

export interface GameState {
    userStore: UserStore;
    state: GameRunningState;
    board?: Board;
    stock?: TileStock;
    activePlayerId: string;
    activePlayerName: string;
}

export interface Game {
    addPlayer: (name: string) => string;
    getGameStatus: (id?: string) => GameStatus;
    startGame: (id: string) => void;
    checkMove: (id: string, tileToMove: TileData) => boolean;
    nextTurn: (id: string) => void;
}

const game = (): Game => {

    const gameState: GameState = {
        userStore: createUserStore(),
        state: GameRunningState.init,
        activePlayerId: "",
        activePlayerName: ""
    }

    const addPlayer = (name: string) => {
        const id = gameState.userStore.createUserSession(name);
        return id;
    }

    const checkMoveInternal = (id: string, tileToPlay: TileData): boolean => {

        const tileOnBoard = gameState.board!.getTileForCoordinates(tileToPlay);
        const moveAllowed = checkMove(gameState, id, tileToPlay);

        console.log(`Tile on board: ${tileToString(tileOnBoard)}`);
        if (moveAllowed) {
            gameState.userStore.moveTileToActiveTurnTiles(id, tileToPlay);
            gameState.board!.placeTileAtBoard(tileToPlay);
            gameState.board!.resizeBoard(tileOnBoard);
        }
        return moveAllowed;
    }

    return {
        addPlayer,
        getGameStatus: (id?) => getGameStatus(gameState, id),
        startGame: (id?) => startGame(gameState, id),
        checkMove: checkMoveInternal,
        nextTurn: (id) => nextTurn(gameState, id)
    };
};

export { game };