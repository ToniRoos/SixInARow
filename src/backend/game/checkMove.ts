import { checkMoveForAlreadyPlayedTilesOfTurn } from "./checkMoveForAlreadyPlayedTilesOfTurn";
import { TileData, TilePosition } from "../../types";
import { GameState } from "../game";
import { findMatchingTilesByCheckingNeighbours } from "../board/findMatchingTilesByCheckingNeighbours";

const checkMove = (gameState: GameState, id: string, tileToMove: TileData): boolean => {

    const board = gameState.board!;
    const tileOnBoard = board.getTileForCoordinates(tileToMove);
    const tilesAtBoard = board.getBoardData().tiles;

    const tileMatches = findMatchingTilesByCheckingNeighbours(tileOnBoard.position, tileToMove.symbol?.color, tileToMove.symbol?.symbol, tilesAtBoard);
    if (!tileMatches) {
        return false;
    }

    let sessionData = gameState.userStore.getSession(id);
    let tilesOnTurn = sessionData ? sessionData.tilesOnTurn : [];

    const tilesOnTurnPosition: TilePosition[] = tilesOnTurn.map(tile => ({ ...tile.position }));
    const allowMove = checkMoveForAlreadyPlayedTilesOfTurn(tilesOnTurnPosition, tileOnBoard.position);

    // if (allowMove && tilesOnTurn.length === 0) {
    //     // const allowMoveForRow = tilesAtBoard.filter(tile => tile.x = tileToMove.x && Math.abs(tile.y - tileToMove.y) === 1)
    // }
    return allowMove;
}

export { checkMove };