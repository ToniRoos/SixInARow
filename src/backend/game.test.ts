
import { game } from "./game";
import { TileSymbols } from "../types";

describe('Board game', () => {

  it('create board', () => {

    const gameInstance = game();
    const player1 = gameInstance.addPlayer('Player 1');
    const player2 = gameInstance.addPlayer('Player 2');

    // ACT
    gameInstance.startGame(player1);
    const gameStatus = gameInstance.getGameStatus();

    expect(gameStatus.board).toMatchSnapshot();
  });

  it('add player', () => {

    const gameInstance = game();
    const player1 = gameInstance.addPlayer('Player 1');

    // ACT
    const gameStatus = gameInstance.getGameStatus();

    expect(gameStatus.players[0]).toBe('Player 1');
  });

  it('add 4 players', () => {

    const gameInstance = game();
    gameInstance.addPlayer('Player 1');
    gameInstance.addPlayer('Player 2');
    gameInstance.addPlayer('Player 3');
    gameInstance.addPlayer('Player 4');

    // ACT
    const gameStatus = gameInstance.getGameStatus();

    expect(gameStatus.players.length).toBe(4);
  });

  it('active player should be player 4', () => {

    const gameInstance = game();
    gameInstance.addPlayer('Player 1');
    gameInstance.addPlayer('Player 2');
    gameInstance.addPlayer('Player 3');
    const playerId = gameInstance.addPlayer('Player 4');

    // ACT
    gameInstance.startGame(playerId);
    const gameStatus = gameInstance.getGameStatus(playerId);

    expect(gameStatus.activeTurn).toBeTruthy();
  });

  it('active player should be player 4', () => {

    const gameInstance = game();
    gameInstance.addPlayer('Player 1');
    gameInstance.addPlayer('Player 2');
    gameInstance.addPlayer('Player 3');
    const playerId = gameInstance.addPlayer('Player 4');

    // ACT
    gameInstance.startGame(playerId);
    const gameStatus = gameInstance.getGameStatus(playerId);

    expect(gameStatus.activeTurn).toBeTruthy();
  });

  it('all 4 players have initial tiles on hand', () => {

    const gameInstance = game();
    const playerId1 = gameInstance.addPlayer('Player 1');
    const playerId2 = gameInstance.addPlayer('Player 2');
    const playerId3 = gameInstance.addPlayer('Player 3');
    const playerId4 = gameInstance.addPlayer('Player 4');

    // ACT
    gameInstance.startGame(playerId4);
    const gameStatusForPlayer1 = gameInstance.getGameStatus(playerId1);
    const gameStatusForPlayer2 = gameInstance.getGameStatus(playerId2);
    const gameStatusForPlayer3 = gameInstance.getGameStatus(playerId3);
    const gameStatusForPlayer4 = gameInstance.getGameStatus(playerId4);

    expect(gameStatusForPlayer1.tilesOnHand.length).toBe(6);
    expect(gameStatusForPlayer2.tilesOnHand.length).toBe(6);
    expect(gameStatusForPlayer3.tilesOnHand.length).toBe(6);
    expect(gameStatusForPlayer4.tilesOnHand.length).toBe(6);
  });

  it('player 1 plays one tile', () => {

    const gameInstance = game();
    const playerId1 = gameInstance.addPlayer('Player 1');
    const playerId2 = gameInstance.addPlayer('Player 2');

    gameInstance.startGame(playerId1);
    const gameStatusForPlayer1 = gameInstance.getGameStatus(playerId1);
    const firstTileToPlay: TileSymbols = gameStatusForPlayer1.tilesOnHand[0];

    // ACT
    gameInstance.checkMove(playerId1, { position: { x: 1, y: 1 }, symbol: firstTileToPlay });
    const gameStatusAfterPlayedTile = gameInstance.getGameStatus(playerId1);

    const boardTile_X1_Y1 = gameStatusAfterPlayedTile.board!.tiles.find(tile => tile.position.x === 1 && tile.position.y === 1);
    expect(boardTile_X1_Y1!.symbol!.color).toBe(firstTileToPlay.color);
    expect(boardTile_X1_Y1!.symbol!.symbol).toBe(firstTileToPlay.symbol);
  });
});