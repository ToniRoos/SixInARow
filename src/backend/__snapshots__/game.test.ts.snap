// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board game create board 1`] = `
Object {
  "activePlayer": "",
  "sizeX": 3,
  "sizeY": 3,
  "tiles": Array [
    Object {
      "position": Object {
        "x": 0,
        "y": 0,
      },
    },
    Object {
      "position": Object {
        "x": 1,
        "y": 0,
      },
    },
    Object {
      "position": Object {
        "x": 2,
        "y": 0,
      },
    },
    Object {
      "position": Object {
        "x": 0,
        "y": 1,
      },
    },
    Object {
      "position": Object {
        "x": 1,
        "y": 1,
      },
    },
    Object {
      "position": Object {
        "x": 2,
        "y": 1,
      },
    },
    Object {
      "position": Object {
        "x": 0,
        "y": 2,
      },
    },
    Object {
      "position": Object {
        "x": 1,
        "y": 2,
      },
    },
    Object {
      "position": Object {
        "x": 2,
        "y": 2,
      },
    },
  ],
}
`;
